tactics/tactics.ml

```diff
@@ -4395,23 +4394,30 @@ let compute_scheme_signature evd scheme names_info ind_type_guess =
   in
   Array.of_list (find_branches 0 (List.rev scheme.branches))
 
-let compute_case_signature env evd mind case names_info =
-  let open Context.Rel.Declaration in
+let compute_case_signature env mind dep names_info =
   let indref = GlobRef.IndRef mind in
-  let branches = case.case_branches in
-  let rec check_branch c = match EConstr.kind evd c with
+  let rec check_branch c = match Constr.kind c with
   | Prod (_,t,c) ->
-    let hd, _ = decompose_app evd t in
+    let hd, _ = Constr.decompose_app t in
     (* no recursive call in case analysis *)
-    let arg = if EConstr.isRefX env evd indref hd then RecArg else OtherArg in
-    (arg, true, not (Vars.noccurn evd 1 c)) :: check_branch c
+    let arg = if Constr.isRefX indref hd then RecArg else OtherArg in
+    (arg, true, not (CVars.noccurn 1 c)) :: check_branch c
   | LetIn (_,_,_,c) ->
-    (OtherArg, false, not (Vars.noccurn evd 1 c)) :: check_branch c
+    (OtherArg, false, not (CVars.noccurn 1 c)) :: check_branch c
   | _ -> []
   in
-  let find_branches lbrch = match lbrch with
-  | LocalAssum (_, t) ->
-    let lchck_brch = check_branch t in
+  let (mib, mip) = Inductive.lookup_mind_specif env mind in
+  let find_branches k =
+    let (ctx, typ) = mip.mind_nf_lc.(k) in
+    let argctx = List.firstn mip.mind_consnrealdecls.(k) ctx in
+    let _, args = Constr.decompose_appvect typ in
+    let _, indices = Array.chop mib.mind_nparams args in
+    let base =
+      if dep then Array.append indices (Context.Rel.instance Constr.mkRel 0 argctx)
+      else indices
+    in
+    let base = Constr.mkApp (Constr.mkProp, base) in (* only used for rel occurrence *)
+    let lchck_brch = check_branch (Term.it_mkProd_or_LetIn base argctx) in
```

Comment:
WTF
can't we modify check_branch to work directly with the context and `base` array instead of producing garbage terms?

Reply:
I guess it's not completely trivial without writing some annoying boilerplate
I think this comment could be clearer though
```suggestion
    let base = Constr.mkApp (Constr.mkProp, base) in (* only used through noccurn *)
    let lchck_brch = check_branch (Term.it_mkProd_or_LetIn base argctx) in
```
"for rel occurence" is IMO slightly ambiguous as it could fit the pattern where we substitute a non-occurring variable with Prop (at least that's what I originally thought it meant when I read the comment)